File: packages/firebase_remote_config/android/src/main/java/io/flutter/plugins/firebase/firebaseremoteconfig/FirebaseRemoteConfigPlugin.java
package io.flutter.plugins.firebase.firebaseremoteconfig;

import android.content.Context;
import android.content.SharedPreferences;
import androidx.annotation.NonNull;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.remoteconfig.FirebaseRemoteConfig;
import com.google.firebase.remoteconfig.FirebaseRemoteConfigFetchThrottledException;
import com.google.firebase.remoteconfig.FirebaseRemoteConfigInfo;
import com.google.firebase.remoteconfig.FirebaseRemoteConfigSettings;
import com.google.firebase.remoteconfig.FirebaseRemoteConfigValue;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/** FirebaseRemoteConfigPlugin */
public class FirebaseRemoteConfigPlugin implements MethodCallHandler {

  public static final String TAG = "FirebaseRemoteConfigPlugin";
  public static final String PREFS_NAME =
      "io.flutter.plugins.firebase.firebaseremoteconfig.FirebaseRemoteConfigPlugin";
  public static final String DEFAULT_PREF_KEY = "default_keys";

  private static SharedPreferences sharedPreferences;
  private final MethodChannel channel;

  public static void registerWith(Registrar registrar) {
    final MethodChannel channel =
        new MethodChannel(registrar.messenger(), "plugins.flutter.io/firebase_remote_config");
    channel.setMethodCallHandler(new FirebaseRemoteConfigPlugin(channel));
    sharedPreferences = registrar.context().getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
  }

  private FirebaseRemoteConfigPlugin(MethodChannel channel) {
    this.channel = channel;
  }

  @Override
  public void onMethodCall(MethodCall call, final Result result) {
    switch (call.method) {
      case "RemoteConfig#instance":
        {
          FirebaseRemoteConfigInfo firebaseRemoteConfigInfo =
              FirebaseRemoteConfig.getInstance().getInfo();

          Map<String, Object> properties = new HashMap<>();
          properties.put("lastFetchTime", firebaseRemoteConfigInfo.getFetchTimeMillis());
          properties.put(
              "lastFetchStatus", mapLastFetchStatus(firebaseRemoteConfigInfo.getLastFetchStatus()));
          properties.put(
              "inDebugMode", firebaseRemoteConfigInfo.getConfigSettings().isDeveloperModeEnabled());
          properties.put("parameters", getConfigParameters());
          result.success(properties);
          break;
        }
      case "RemoteConfig#setConfigSettings":
        {
          boolean debugMode = call.argument("debugMode");
          final FirebaseRemoteConfig firebaseRemoteConfig = FirebaseRemoteConfig.getInstance();
          FirebaseRemoteConfigSettings settings =
              new FirebaseRemoteConfigSettings.Builder().setDeveloperModeEnabled(debugMode).build();
          firebaseRemoteConfig.setConfigSettings(settings);
          result.success(null);
          break;
        }
      case "RemoteConfig#fetch":
        {
          long expiration = ((Number) call.argument("expiration")).longValue();
          final FirebaseRemoteConfig firebaseRemoteConfig = FirebaseRemoteConfig.getInstance();
          firebaseRemoteConfig
              .fetch(expiration)
              .addOnCompleteListener(
                  new OnCompleteListener<Void>() {
                    @Override
                    public void onComplete(@NonNull Task<Void> task) {
                      FirebaseRemoteConfigInfo firebaseRemoteConfigInfo =
                          firebaseRemoteConfig.getInfo();
                      Map<String, Object> properties = new HashMap<>();
                      properties.put(
                          "lastFetchTime", firebaseRemoteConfigInfo.getFetchTimeMillis());
                      properties.put(
                          "lastFetchStatus",
                          mapLastFetchStatus(firebaseRemoteConfigInfo.getLastFetchStatus()));
                      if (!task.isSuccessful()) {
                        final Exception exception = task.getException();

                        if (exception instanceof FirebaseRemoteConfigFetchThrottledException) {
                          properties.put(
                              "fetchThrottledEnd",
                              ((FirebaseRemoteConfigFetchThrottledException) exception)
                                  .getThrottleEndTimeMillis());
                          String errorMessage =
                              "Fetch has been throttled. See the error's "
                                  + "FETCH_THROTTLED_END field for throttle end time.";
                          result.error("fetchFailedThrottled", errorMessage, properties);
                        } else {
                          String errorMessage =
                              "Unable to complete fetch. Reason is unknown "
                                  + "but this could be due to lack of connectivity.";
                          result.error("fetchFailed", errorMessage, properties);
                        }
                      } else {
                        result.success(properties);
                      }
                    }
                  });
          break;
        }
      case "RemoteConfig#activate":
        {
          boolean newConfig = FirebaseRemoteConfig.getInstance().activateFetched();
          Map<String, Object> properties = new HashMap<>();
          properties.put("parameters", getConfigParameters());
          properties.put("newConfig", newConfig);
          result.success(properties);
          break;
        }
      case "RemoteConfig#setDefaults":
        {
          Map<String, Object> defaults = call.argument("defaults");
          FirebaseRemoteConfig.getInstance().setDefaults(defaults);
          SharedPreferences.Editor editor = sharedPreferences.edit();
          editor.putStringSet(DEFAULT_PREF_KEY, defaults.keySet()).apply();
          result.success(null);
          break;
        }
      default:
        {
          result.notImplemented();
          break;
        }
    }
  }

  private Map<String, Object> getConfigParameters() {
    FirebaseRemoteConfig firebaseRemoteConfig = FirebaseRemoteConfig.getInstance();
    Map<String, Object> parameterMap = new HashMap<>();
    Set<String> keys = firebaseRemoteConfig.getKeysByPrefix("");
    for (String key : keys) {
      FirebaseRemoteConfigValue remoteConfigValue = firebaseRemoteConfig.getValue(key);
      parameterMap.put(key, createRemoteConfigValueMap(remoteConfigValue));
    }
    // Add default parameters if missing since `getKeysByPrefix` does not return default keys.
    Set<String> defaultKeys =
        sharedPreferences.getStringSet(DEFAULT_PREF_KEY, new HashSet<String>());
    for (String defaultKey : defaultKeys) {
      if (!parameterMap.containsKey(defaultKey)) {
        FirebaseRemoteConfigValue remoteConfigValue = firebaseRemoteConfig.getValue(defaultKey);
        parameterMap.put(defaultKey, createRemoteConfigValueMap(remoteConfigValue));
      }
    }
    return parameterMap;
  }

  private Map<String, Object> createRemoteConfigValueMap(
      FirebaseRemoteConfigValue remoteConfigValue) {
    Map<String, Object> valueMap = new HashMap<>();
    valueMap.put("value", remoteConfigValue.asByteArray());
    valueMap.put("source", mapValueSource(remoteConfigValue.getSource()));
    return valueMap;
  }

  private String mapLastFetchStatus(int status) {
    switch (status) {
      case FirebaseRemoteConfig.LAST_FETCH_STATUS_SUCCESS:
        return "success";
      case FirebaseRemoteConfig.LAST_FETCH_STATUS_FAILURE:
        return "failure";
      case FirebaseRemoteConfig.LAST_FETCH_STATUS_THROTTLED:
        return "throttled";
      case FirebaseRemoteConfig.LAST_FETCH_STATUS_NO_FETCH_YET:
        return "noFetchYet";
      default:
        return "failure";
    }
  }

  private String mapValueSource(int source) {
    switch (source) {
      case FirebaseRemoteConfig.VALUE_SOURCE_STATIC:
        return "static";
      case FirebaseRemoteConfig.VALUE_SOURCE_DEFAULT:
        return "default";
      case FirebaseRemoteConfig.VALUE_SOURCE_REMOTE:
        return "remote";
      default:
        return "static";
    }
  }
}

File: packages/firebase_remote_config/ios/Classes/FirebaseRemoteConfigPlugin.m
#import "FirebaseRemoteConfigPlugin.h"

#import <Firebase/Firebase.h>

@interface FirebaseRemoteConfigPlugin ()
@property(nonatomic, retain) FlutterMethodChannel *channel;
@end

@implementation FirebaseRemoteConfigPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
  FlutterMethodChannel *channel =
      [FlutterMethodChannel methodChannelWithName:@"plugins.flutter.io/firebase_remote_config"
                                  binaryMessenger:[registrar messenger]];
  FirebaseRemoteConfigPlugin *instance = [[FirebaseRemoteConfigPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (instancetype)init {
  self = [super init];
  if (self) {
    if (![FIRApp appNamed:@"__FIRAPP_DEFAULT"]) {
      NSLog(@"Configuring the default Firebase app...");
      [FIRApp configure];
      NSLog(@"Configured the default Firebase app %@.", [FIRApp defaultApp].name);
    }
  }
  return self;
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
  if ([@"RemoteConfig#instance" isEqualToString:call.method]) {
    FIRRemoteConfig *remoteConfig = [FIRRemoteConfig remoteConfig];
    FIRRemoteConfigSettings *firRemoteConfigSettings = [remoteConfig configSettings];
    NSMutableDictionary *resultDict = [[NSMutableDictionary alloc] init];

    resultDict[@"lastFetchTime"] = [[NSNumber alloc]
        initWithLong:(long)[[remoteConfig lastFetchTime] timeIntervalSince1970] * 1000];
    resultDict[@"lastFetchStatus"] =
        [self mapLastFetchStatus:(FIRRemoteConfigFetchStatus)[remoteConfig lastFetchStatus]];
    resultDict[@"inDebugMode"] =
        [[NSNumber alloc] initWithBool:[firRemoteConfigSettings isDeveloperModeEnabled]];

    resultDict[@"parameters"] = [self getConfigParameters];

    result(resultDict);
  } else if ([@"RemoteConfig#setConfigSettings" isEqualToString:call.method]) {
    FIRRemoteConfig *remoteConfig = [FIRRemoteConfig remoteConfig];
    bool debugMode = (bool)call.arguments[@"debugMode"];
    FIRRemoteConfigSettings *remoteConfigSettings =
        [[FIRRemoteConfigSettings alloc] initWithDeveloperModeEnabled:debugMode];
    [remoteConfig setConfigSettings:remoteConfigSettings];
    result(nil);
  } else if ([@"RemoteConfig#fetch" isEqualToString:call.method]) {
    FIRRemoteConfig *remoteConfig = [FIRRemoteConfig remoteConfig];
    long expiration = (long)call.arguments[@"expiration"];

    [remoteConfig
        fetchWithExpirationDuration:expiration
                  completionHandler:^(FIRRemoteConfigFetchStatus status, NSError *error) {
                    NSNumber *lastFetchTime = [[NSNumber alloc]
                        initWithLong:(long)[[remoteConfig lastFetchTime] timeIntervalSince1970] *
                                     1000];
                    NSString *lastFetchStatus =
                        [self mapLastFetchStatus:(FIRRemoteConfigFetchStatus)[remoteConfig
                                                                                  lastFetchStatus]];
                    NSMutableDictionary *resultDict = [[NSMutableDictionary alloc] init];
                    resultDict[@"lastFetchTime"] = lastFetchTime;
                    resultDict[@"lastFetchStatus"] = lastFetchStatus;

                    if (status != FIRRemoteConfigFetchStatusSuccess) {
                      FlutterError *flutterError;
                      if (status == FIRRemoteConfigFetchStatusThrottled) {
                        int mills =
                            [[error.userInfo
                                valueForKey:FIRRemoteConfigThrottledEndTimeInSecondsKey] intValue] *
                            1000;
                        resultDict[@"fetchThrottledEnd"] = [[NSNumber alloc] initWithInt:mills];
                        NSString *errorMessage =
                            @"Fetch has been throttled. See the error's fetchThrottledEnd "
                             "field for throttle end time.";
                        flutterError = [FlutterError errorWithCode:@"fetchFailedThrottled"
                                                           message:errorMessage
                                                           details:resultDict];
                      } else {
                        NSString *errorMessage = @"Unable to complete fetch. Reason is unknown "
                                                  "but this could be due to lack of connectivity.";
                        flutterError = [FlutterError errorWithCode:@"fetchFailed"
                                                           message:errorMessage
                                                           details:resultDict];
                      }
                      result(flutterError);
                    } else {
                      result(resultDict);
                    }
                  }];
  } else if ([@"RemoteConfig#activate" isEqualToString:call.method]) {
    BOOL newConfig = [[FIRRemoteConfig remoteConfig] activateFetched];
    NSDictionary *parameters = [self getConfigParameters];
    NSMutableDictionary *resultDict = [[NSMutableDictionary alloc] init];
    resultDict[@"newConfig"] = [NSNumber numberWithBool:newConfig];
    resultDict[@"parameters"] = parameters;
    result(resultDict);
  } else if ([@"RemoteConfig#setDefaults" isEqualToString:call.method]) {
    FIRRemoteConfig *remoteConfig = [FIRRemoteConfig remoteConfig];
    NSDictionary *defaults = call.arguments[@"defaults"];
    [remoteConfig setDefaults:defaults];
    result(nil);
  } else {
    result(FlutterMethodNotImplemented);
  }
}

- (NSMutableDictionary *)createRemoteConfigValueDict:(FIRRemoteConfigValue *)remoteConfigValue {
  NSMutableDictionary *valueDict = [[NSMutableDictionary alloc] init];
  valueDict[@"value"] = [FlutterStandardTypedData typedDataWithBytes:[remoteConfigValue dataValue]];
  valueDict[@"source"] = [self mapValueSource:[remoteConfigValue source]];
  return valueDict;
}

- (NSDictionary *)getConfigParameters {
  FIRRemoteConfig *remoteConfig = [FIRRemoteConfig remoteConfig];
  NSMutableDictionary *parameterDict = [[NSMutableDictionary alloc] init];
  NSSet *keySet = [remoteConfig keysWithPrefix:@""];
  for (NSString *key in keySet) {
    parameterDict[key] = [self createRemoteConfigValueDict:[remoteConfig configValueForKey:key]];
  }
  // Add default parameters if missing since `keysWithPrefix` does not return default keys.
  NSArray *defaultKeys = [remoteConfig allKeysFromSource:FIRRemoteConfigSourceDefault
                                               namespace:FIRNamespaceGoogleMobilePlatform];
  for (NSString *key in defaultKeys) {
    if ([parameterDict valueForKey:key] == nil) {
      parameterDict[key] = [self createRemoteConfigValueDict:[remoteConfig configValueForKey:key]];
    }
  }
  return parameterDict;
}

- (NSString *)mapLastFetchStatus:(FIRRemoteConfigFetchStatus)status {
  if (status == FIRRemoteConfigFetchStatusSuccess) {
    return @"success";
  } else if (status == FIRRemoteConfigFetchStatusFailure) {
    return @"failure";
  } else if (status == FIRRemoteConfigFetchStatusThrottled) {
    return @"throttled";
  } else if (status == FIRRemoteConfigFetchStatusNoFetchYet) {
    return @"noFetchYet";
  } else {
    return @"failure";
  }
}

- (NSString *)mapValueSource:(FIRRemoteConfigSource)source {
  if (source == FIRRemoteConfigSourceStatic) {
    return @"static";
  } else if (source == FIRRemoteConfigSourceDefault) {
    return @"default";
  } else if (source == FIRRemoteConfigSourceRemote) {
    return @"remote";
  } else {
    return @"static";
  }
}

@end
